jujugui/static/gui/src/app/store/env/bakery.js

```diff
@@ -0,0 +1,339 @@
+       @param {Object} an exported Macaroon
+       */
+      _setMacaroonsCookie: function (originalRequest, jsonMacaroon) {
+        Y.Cookie.set('Macaroons', btoa(JSON.stringify(jsonMacaroon)));
+        originalRequest();
+      },
+
+      /**
+       Go to the discharge endpoint to obtain the third party discharge.
+
+       @method obtainThirdPartyDischarge
+       @param {String} The origin location
+       @param {String} The third party location where to discharge
+       @param {Function} The macaroon to be discharge
+       @param {Function} The request to be sent again in case of
+       successful authentication
+       @param {Function} The callback failure in case of wrong authentication
+       */
+      _obtainThirdPartyDischarge: function (location,
+                                            thirdPartyLocation, condition,
+                                            successCallback, failureCallback) {
+        thirdPartyLocation += '/discharge';
+        var headers = {
+          'Bakery-Protocol-Version': 1,
+          'Content-Type': 'application/x-www-form-urlencoded'
+        };
+        var content = 'id=' + encodeURI(condition) +
+          '&location=' + encodeURI(location);
+        this.webhandler.sendPostRequest(
+          thirdPartyLocation,
+          headers, content, null, null, false, null,
+          this._requestHandler.bind(
+            this,
+            this._exportMacaroon.bind(this, successCallback, failureCallback),
+            this._interact.bind(this, successCallback, failureCallback)
+          )
+        );
+      },
+
+    /**
+      Get a JSON response from authentication either trusted or with
+      interaction that contains a macaroon.
+
+      @method _exportMacaroon
+      @param {Function} The callback function to be sent in case of
+                        successful authentication
+      @param {Function} The callback function failure in case of
+                        wrong authentication
+      @param {Object} response The XHR response object.
+    */
+    _exportMacaroon: function (successCallback, failureCallback, response) {
+      try {
+        var dm = macaroon.import(
+          JSON.parse(response.target.responseText).Macaroon
+        );
+        successCallback(dm);
+      } catch (ex) {
+        failureCallback(ex.message);
+      }
+    },
+
+    /**
+      Interact to be able to sign-in to get authenticated.
+
+      @method _interact
+      @param {Function} The callback function to be sent in case of
+                        successful authentication
+      @param {Function} The callback function failure in case of
+                        wrong authentication
+    */
+    _interact: function(successCallback, failureCallback, e) {
+      var response = JSON.parse(e.target.responseText);
+      if (response.Code !== 'interaction required') {
+        failureCallback(response.Code);
+        return;
+      }
+
+      window.open(response.Info.VisitURL, 'Login');
```

Comment:
For custom login methods, like the horizon login there will need to be the ability to use custom visit functions here.